packages/next/shared/lib/match-remote-pattern.ts

```diff
@@ -0,0 +1,68 @@
+import type { RemotePattern } from './image-config'
+
+export function matchRemotePattern(pattern: RemotePattern, url: URL): boolean {
+  if (pattern.protocol !== undefined) {
+    const actualProto = url.protocol.slice(0, -1)
+    if (pattern.protocol !== actualProto) {
+      return false
+    }
+  }
+  if (pattern.port !== undefined) {
+    if (pattern.port !== url.port) {
+      return false
+    }
+  }
+  if (pattern.pathname !== undefined) {
+    const patternParts = pattern.pathname.split('/')
+    const actualParts = url.pathname.split('/')
+    const len = Math.max(patternParts.length, actualParts.length)
+    for (let i = 0; i < len; i++) {
+      if (patternParts[i] === '**' && actualParts[i] !== undefined) {
+        // Double asterisk means "match everything until the end of the path"
+        // so we can break the loop early
+        break
+      }
+      if (patternParts[i] === '*') {
+        // Single asterisk means "match this part" so we can
+        // continue to the next part of the loop
+        continue
+      }
+      if (patternParts[i] !== actualParts[i]) {
+        return false
+      }
+    }
+  }
+
+  if (pattern.hostname !== undefined) {
```

Comment:
Should this error if `hostname` is missing as a safe guard?

Reply:
Added in 8d9e8fa